metadata/src/main/java/org/apache/kafka/image/ScramImage.java

```diff
@@ -0,0 +1,97 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.image;
+
+import org.apache.kafka.image.writer.ImageWriter;
+import org.apache.kafka.image.writer.ImageWriterOptions;
+import org.apache.kafka.clients.admin.ScramMechanism;
+
+// import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+// import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+
+/**
+ * Represents the SCRAM credentials in the metadata image.
+ *
+ * This class is thread-safe.
+ */
+public final class ScramImage {
+    public static final ScramImage EMPTY = new ScramImage(Collections.emptyMap());
+
+    private final Map<ScramMechanism, Map<String, ScramCredentialData>> mechanisms;
+
+    public ScramImage(Map<ScramMechanism, Map<String, ScramCredentialData>> mechanisms) {
+        this.mechanisms = Collections.unmodifiableMap(mechanisms);
+    }
+
+    public void write(ImageWriter writer, ImageWriterOptions options) {
+        for (Entry<ScramMechanism, Map<String, ScramCredentialData>> mechanismEntry : mechanisms.entrySet()) {
+            for (Entry<String, ScramCredentialData> userEntry : mechanismEntry.getValue().entrySet()) {
+                writer.write(0, userEntry.getValue().toRecord(userEntry.getKey(), mechanismEntry.getKey()));
+            }
+        }
+    }
+
+    public Map<ScramMechanism, Map<String, ScramCredentialData>> mechanisms() {
+        return mechanisms;
+    }
+
+    public boolean isEmpty() {
+        return mechanisms.isEmpty();
+    }
+
+    @Override
+    public int hashCode() {
+        return mechanisms.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null) return false;
+        if (!o.getClass().equals(ScramImage.class)) return false;
+        ScramImage other = (ScramImage) o;
+        return mechanisms.equals(other.mechanisms);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
+        builder.append("ScramImage(");
+        List<ScramMechanism> sortedMechanisms = mechanisms.keySet().stream().sorted().collect(Collectors.toList());
+        String preMechanismComma = "";
+        for (ScramMechanism mechanism : sortedMechanisms) {
+            builder.append(preMechanismComma).append(mechanism).append(": {");
+            Map<String, ScramCredentialData> userMap = mechanisms.get(mechanism);
+            List<String> sortedUserNames = userMap.keySet().stream().sorted().collect(Collectors.toList());
+            String preUserNameComma = "";
+            for (String userName : sortedUserNames) {
+                builder.append(preUserNameComma).append(userName).append("=").append(userMap.get(userName));
```

Comment:
we should not print  sensitive info like `userMap.get(`

Reply:
As far as I can tell, the toString method is never used.
I think it is useful for debugging, but I'd like to know how we handle this in general.

Reply:
This one should b okay. It is just printing a list of usernames and which SCRAM mechanism they use. No passwords or salts displayed here.

Reply:
Fixed